### PR TITLE
Browser safe list support

### PR DESCRIPTION
**Pull request by heidijinxujia**
**Problem**
To avoid the broken user experience when auth through browsers.

**Proposal**
BrowserDescriptor describes the browser, including the package name, browser's signature hashes, if using the custom tab and the version range that supports the MSAL browser authorization flow.
```json
{
  "browser_package_name": "com.android.chrome",
  "browser_signature_hashes": [
 "7fmduHKTdHHrlMvldlEqAIlSfii1tl35bxj1OXN5Ve8c4lU6URVu4xtSHc3BVZxS6WWJnxMDhIfQN0N0K2NDJg=="
      ],
  "browser_use_customTab" : true,
  "browser_version_lower_bound" : "45"
}
```

The browser safe list is a list of BrowserDescriptor objects and is defined in the msal configuration json file. Before MSAL app trying to launch a browser, it will look into the browser safe list. If there is a match, the app will launch the matching browser, otherwise, MSAL app will try the next installed browser until it finds a match. If no matched browser found, MSAL will fallback into the embedded webview flow. However, for the B2C scenario, as the browser flow is required, if no matching browser found, MSAL will throw the ClientException to the dev.

**Work Items**
- Add the BrowserDescriptor class to represent the browser attributes
- Add the BrowserSafeList property into the BrowserAuthorizationStrategy
- Add the safe list filtering inside the browser selecting logic during the authorization process
- Add unit test

**Discussion**
For the safe list object, is it better to keep it inside the BrowserAuthorizationStrategy or make it a single instance class?

### CODE REVIEW

File: common/src/main/java/com/microsoft/identity/common/internal/ui/browser/BrowserDescriptor.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    /**
+     * The function to compare the two versions.
+     *
+     * @param thisVersion
+     * @param thatVersion
+     * @return int -1 if thisVersion is smaller than thatVersion,
+     *         1 if thisVersion is larger than thatVersion,
+     *         0 if thisVersion is equal to thatVersion.
+     */
+    private int compareSemanticVersion(
```
Comment (iambmelt) [inline common/src/main/java/com/microsoft/identity/common/internal/ui/browser/BrowserDescriptor.java]: Didn't we just add this function somewhere else? For broker `hello()`?

Comment (iambmelt) [reply common/src/main/java/com/microsoft/identity/common/internal/ui/browser/BrowserDescriptor.java]: `BrokerUtils` -- this should be refactored into `common`, not copy/pasted

Comment (heidijinxujia) [reply common/src/main/java/com/microsoft/identity/common/internal/ui/browser/BrowserDescriptor.java]: Thank you, Brian. Addressed.

File: common/src/androidTest/java/com/microsoft/identity/common/BrowserSelectorTest.java [hunk 2]
```diff
@@ -0,0 +1,221 @@
+package com.microsoft.identity.common;
```
Comment (iambmelt) [inline common/src/androidTest/java/com/microsoft/identity/common/BrowserSelectorTest.java]: License info

**Comment by iambmelt** [comment]
There is a failing unit test

File: common/src/main/java/com/microsoft/identity/common/internal/ui/browser/BrowserSelector.java [hunk 3]
```diff
@@ -48,21 +48,31 @@
...
-            return allBrowsers.get(0);
-        } else {
-            Logger.error(TAG, "No available browser installed on the device.", null);
-            throw new ClientException(ErrorStrings.NO_AVAILABLE_BROWSER_FOUND, "No available browser installed on the device.");
+
+        for (Browser browser : allBrowsers) {
+            for (BrowserDescriptor browserDescriptor : browserSafeList) {
+                if (browserDescriptor.matches(browser)) {
+                    Logger.verbose(TAG, "Select the browser to launch.");
+                    Logger.verbosePII(
```
Comment (iambmelt) [inline common/src/main/java/com/microsoft/identity/common/internal/ui/browser/BrowserSelector.java]: I think we can probably log this as non-PII without issue - is there something I'm not considering?

**Pull request merged**